### PR TITLE
Fix headers guards

### PR DESCRIPTION
**Pull request by ceolin**
Identifiers starting with an underscore followed by either uppercase letter or other underscore are reserved and should not be used according with C99.

This pr address this problem on headers guards only. #9882

### CODE REVIEW

**Review by jhedberg — APPROVED** [review]
Went through the Bluetooth part - looks fine to me

**Comment by ceolin** [comment]
@pfalcon fair enough. Can I prefix every header guard with "HEADER_GUARD" ? or "HEADER" ?

@jhedberg are you ok doing it to bluetooth as well ?

**Comment by pfalcon** [comment]
> @pfalcon fair enough. Can I prefix every header guard with "HEADER_GUARD" ? or "HEADER" ?

Well, you're trying to make Zephyr compliant and well-behaving. Just imagine if some other / all projects would prefix their header guards with something simple and obvious, like "HEADER_GUARD" - they all would be clashing/non-compliant. IMHO, if doing another pass on that at all, then do it right, and that means doing project-specific namespacing. YMMV

**Comment by nashif** [comment]
@ceolin follow this guide for example:

https://google.github.io/styleguide/cppguide.html#The__define_Guard

I would not go all the way and add the path, but ZEPHYR_ADC_H should do it.

I would start a document (or use an existing one) to document those and make it part of the Zephyr style and naming conventions.

**Comment by jhedberg** [comment]
@ceolin IMO having _H in the define already spells sufficiently "header". I'm fine with the current solution, or if you want to add a __ suffix I'm not opposed to that either.

**Comment by pfalcon** [comment]
@nashif :

The doc you link to clearly has a trailing underscore.

> I would not go all the way and add the path, but ZEPHYR_ADC_H should do it.

ZEPHYR_ADC_H means "something high (e.g. high part) of ADC value in Zephyr project (where ZEPHYR is ~~mandatory~~ namespacing prefix for all API symbols). Again, perfect name for a public symbol, not header guard.

**Comment by nashif** [comment]
> The doc you link to clearly has a trailing underscore.

yes, I forgot the trailing underscore, ZEPHYR_ADC_H_ is fine, but I can also live with ADC_H_, no preferences really. The only issue with ZEPHYR_ADC_H_, it is long.

**Comment by ceolin** [comment]
A trailing underscore should be enough to mess with our API. Putting the whole namespace will make some guards really big, I changed the guards within kernel/* because are few of them and are my main focus now and tried to keep the others (fixing what made them not compliant).

So, can we proceed with these changes adding only a trailing underscore ?

**Comment by pfalcon** [comment]
> The only issue with ZEPHYR_ADC_H_, it is long.

Long is good. The longer, the less probability of clash, pure mathematical statistics.

**Comment by ceolin** [comment]
@pfalcon I've no problem with long names for header guards, this is basically one time thing. We just need to get a conclusion on this. Prefix everything with ZEPHYR and add a trailing underscore resolve our problem, or we can be more strict and use the whole file's path. The latter is a little bit to much IMHO.

**Comment by pfalcon** [comment]
@ceolin : I'm only a developer, so can only give suggestions. You can combine votes from different developers, requests from project leaders, and your own opinion to get to the result.

If the choices are:

1. ZEPHYR_ + filename + H_
2. ZEPHYR_ + pathname + H_

My vote is for p.2, as it's more formal and more unambiguous. If for some reason choice 2 doesn't include "ZEPHYR_", I vote for that choice which does include prefix ;-). Indeed, header guards are write-once, so optimizing them for length doesn't make much sense, instead optimizing for formal uniqueness does, IMHO.

**Comment by nashif** [comment]
i vote for 

`ZEPHYR_ + filename + H_`

**Comment by ceolin** [comment]
I second for option 2, we have files with same name on project.

**Comment by ceolin** [comment]
```
p/zephyr » find -name "util.h"
./drivers/bluetooth/util.h
./include/misc/util.h
./subsys/bluetooth/controller/util/util.h
```

**Review by andyross — APPROVED** [review]
Not a compaint about this patch.  No one cares about include guard symbol names.  This is fine.

That said: the spirit of this "no leading underscores" rule is that those symbols are reserved for internal use by "the system".

And **we are the system** in this context.    Disallowing Zephyr from using this convention (which operating systems have been doing consistently in C since pre-PDP11 unix!) is doing absolutely nothing but adding bytes to code and IMHO making us look like dumb slaves to a ridiculous specification.  But... meh.  Names is names.

**Comment by nashif** [comment]
#WeAreTheSystem

**Comment by nashif** [comment]
@ceolin I am fine with the above.

**Comment by ceolin** [comment]
@andyross As fair as I understood this is reserved for the compiler and not for the system.

@nashif are you fine with the path in the guards ? or is something else ?

**Comment by nashif** [comment]
> @nashif are you fine with the path in the guards ? or is something else ?

path in the guard.
to @andyross point, this is a one time change and we will not bother with that again, what we need to do is however, any change needs to be added to some guide to developers so we can point them to it if wrong style is being added to new changes, can you start with such guide?

**Comment by ceolin** [comment]
@nashif ack !

**Comment by rettichschnidi** [comment]
How about simply using [#pragma once](https://en.wikipedia.org/wiki/Pragma_once)? While it is not in the C standard it does not violate it and every relevant compiler supports it officially.

**Comment by ceolin** [comment]
I rather prefer using `#pragma` too, the problem is that to be MISRA C compliant we have to follow the C standard, and to be honest there are other advantages of doing it too.

**Comment by rettichschnidi** [comment]
MISRA C2012 does allow the usage of extensions (#pragma once in this case) as long as the implementation-defined behavior is documented and understood.

**Comment by ceolin** [comment]
@rettichschnidi there is one advisory guideline that says to not use language extensions. You're right, you still can use if the behaviour is defined and you document it. But to me, is just easier change these guards to be compliant than need to document it for each supported compiler.

I know that we inevitable will need to do it for some guidelines, but we should avoid this for this simple fixes.

File: arch/posix/include/posix_cheats.h [hunk 1]
```diff
@@ -9,8 +9,8 @@
  * Header to be able to compile the Zephyr kernel on top of a POSIX OS
  */
 
-#ifndef _POSIX_CHEATS_H
-#define _POSIX_CHEATS_H
```
Comment (aescolar) [inline arch/posix/include/posix_cheats.h]: You need to search and replace _POSIX_CHEATS_H in the CMake files to (that's why regression is failing)

**Review by aescolar — DISMISSED** [review]

File: arch/posix/include/posix_cheats.h (hunk 1 above)
Comment (aescolar) [reply arch/posix/include/posix_cheats.h]: Or better, use that guard name but replace the 1st line
to `#if !defined(ZEPHYR_ARCH_POSIX_INCLUDE_POSIX_CHEATS_H_) && !defined(NO_POSIX_CHEATS)`
And replace in the CMake files `_POSIX_CHEATS_H` with `NO_POSIX_CHEATS`

Comment (ceolin) [reply arch/posix/include/posix_cheats.h]: @aescolar thanks, I'll fix it

**Comment by ceolin** [comment]
@aescolar can you check it again please ?

File: arch/posix/include/posix_cheats.h (hunk 1 above)
Comment (nashif) [reply arch/posix/include/posix_cheats.h]: should be fixed now...

**Comment by aescolar** [comment]
> @aescolar can you check it again please ?

It seems fine, yes. Thanks

**Review by nashif — APPROVED** [review]

**Pull request merged**